docx/oxml/text/run.py

```diff
@@ -136,36 +136,40 @@ def text(self):
         equivalent.
         """
-        text = ''
-        if CT_FldChar.numOfNestedFldChar > 0:
-            for child in self:
-                if child.tag == qn('w:fldChar'):
-                    if child.fldCharType == 'begin':
-                        CT_FldChar.numOfNestedFldChar += 1
-                    else:
-                        # `w:fldChar` stores instruction text from `w:fldCharType='begin'`
-                        # to `w:fldCharType='separate'` if 'separate' exists, if not then
-                        # until `w:fldCharType='end'`.
-                        CT_FldChar.numOfNestedFldChar -= 1
-        else:
-            for child in self:
-                if child.tag == qn('w:t'):
-                    t_text = child.text
-                    text += t_text if t_text is not None else ''
-                elif child.tag == qn('w:tab'):
-                    text += '\t'
-                elif child.tag == qn('w:br'):
-                    text += '\n'
-                elif child.tag == qn('w:cr'):
-                    text += '\r'
-                elif child.tag == qn('w:noBreakHyphen'):
-                    text += '-'
-                elif child.tag == qn('w:sym'):
-                    text += child.readSymbol
-                elif child.tag == qn('w:fldChar'):
-                    if child.fldCharType == 'begin':
-                        CT_FldChar.numOfNestedFldChar += 1
-        return text
+        text = {'val':''}
+        def appendText(newText):
+            # ignore text if it's in complex field char
+            if CT_FldChar.numOfNestedFldChar == 0:
+                text['val'] += newText
+        for child in self:
+            if child.tag == qn('w:t'):
+                t_text = child.text
+                appendText(t_text if t_text is not None else '')
+            elif child.tag == qn('w:tab'):
+                appendText('\t')
+            elif child.tag == qn('w:br'):
+                appendText('\n')
+            elif child.tag == qn('w:cr'):
+                appendText('\r')
+            elif child.tag == qn('w:noBreakHyphen'):
+                appendText('-')
+            elif child.tag == qn('w:sym'):
+                appendText(child.readSymbol)
+            elif child.tag == qn('w:fldChar'):
+                # `fldChar` should ignore text form types form
+                # `begin` to `end` or from `begin` to `separate`
+                if child.fldCharType == 'begin':
+                    CT_FldChar.numOfNestedFldChar += 1
+                elif child.fldCharType == 'separate':
+                    CT_FldChar.numOfNestedFldChar -= 1
+                    # should ignore the next `end`, because text
+                    # is shown from `separate` to `end`
+                    CT_FldChar.hasSeparate = True
+                elif CT_FldChar.hasSeparate == False:
+                    CT_FldChar.numOfNestedFldChar -= 1
+                else:
+                    CT_FldChar.hasSeparate = False
```

Comment:
It would be nice just to add comment above this code to say: "We know this `fldCharType == 'end'` so we check if this `FldChar` has separate in it.